operator/pkg/apis/infrastructure/v1alpha1/controlplane_defaults.go

```
@@ -0,0 +1,97 @@
+// CRD instance object in Kubernetes. All the defaults are set while reconciling
+// over ControlPlane object in the controllers.
+func (c *ControlPlane) SetDefaults(ctx context.Context) {
+	c.Spec.SetDefaults(ctx)
+}
+
+// SetDefaults for the ControlPlaneSpec, cascading to all subspecs
+func (s *ControlPlaneSpec) SetDefaults(ctx context.Context) {}
+
+// WithStaticDefaults
+func (p *ControlPlaneSpec) WithStaticDefaults(defaultStaticSpec func() *ControlPlaneSpec) (_ ControlPlaneSpec, err error) {
+	userConfig := p.DeepCopy()
+	patchedCPSpec, err := defaultStaticSpec().Patch(userConfig)
+	return patchedCPSpec, err
+}
+
+// Patch will overwrite the receiver with desired fields in the parameter provided
+func (c *ControlPlaneSpec) Patch(patch *ControlPlaneSpec) (spec ControlPlaneSpec, err error) {
+	if patch.KubernetesVersion != "" {
+		c.KubernetesVersion = patch.KubernetesVersion
+	}
+	if c.Etcd, err = patch.Etcd.WithStaticDefaults(context.TODO(), func() *ETCDSpec {
+		return &c.Etcd
+	}); err != nil {
+		return ControlPlaneSpec{}, err
+	}
+	return *c, err
+}
+
+func (s *ETCDSpec) WithStaticDefaults(ctx context.Context, defaultStaticSpec func() *ETCDSpec) (_ ETCDSpec, err error) {
+	spec := s.DeepCopy()
+	patchedEtcdSpec, err := defaultStaticSpec().Patch(spec)
+	return patchedEtcdSpec, err
+}
+
+func (s *ETCDSpec) Patch(patch *ETCDSpec) (ETCDSpec, error) {
+	if patch.Instances.AMI != "" {
+		s.Instances.AMI = patch.Instances.AMI
+	}
+	if patch.Instances.Type != "" {
+		s.Instances.Type = patch.Instances.Type
+	}
+	if patch.Spec != nil {
+		obj := v1.PodSpec{}
+		mergedPatch, err := mergePatch(s.Spec, patch.Spec, obj)
+		if err != nil {
+			return ETCDSpec{}, fmt.Errorf("failed to merge patch, %w", err)
+		}
+		if err := json.Unmarshal(mergedPatch, s.Spec); err != nil {
+			return ETCDSpec{}, fmt.Errorf("unmarshalling mergedPatch to podSpec, %w", err)
+		}
+	}
+	return *s, nil
+}
+
+func mergePatch(defaultObj, patch, object interface{}) ([]byte, error) {
```

Is there a way to guarantee this can't error? It would clean up the code a lot.

---

Tough because we want to return an error to the caller if this fails, not fail silently. 
Other way could be returned []bytes can be nil but doesn't solve the purpose.

---

What I mean is can we guarantee the types all line up so it would be impossible to error? Maybe not, but just wanted to check.